init, test: improve network reachability test coverage and safety

jonatack opened the pull request:
Adds missing network reachability test coverage and an assertion during init, noticed while reviewing #22834:

- assert during init that each network reachability is  true by default
- add CJDNS to the `LimitedAndReachable_Network` unit tests
- hoist proxy out of two network loops in feature_proxy.py
- test that passing invalid `-proxy` raises expected init error
- test that passing invalid `-onion` raises expected init error
- test that passing `-onlynet=onion` without `-proxy` and `-onion` raises expected init error
- test that passing `-onlynet=onion` with `-onion=0` and with `-noonion` raises expected init error

jonatack committed:
net, init: assert each network reachability is true by default

The default network reachability values are implicitly set
by this line in net.cpp:

static bool vfLimited[NET_MAX] GUARDED_BY(g_maplocalhost_mutex) = {};

This commit asserts that each network is reachable during
the first loop through them during bitcoind init.

jonatack committed:
test: add CJDNS to LimitedAndReachable_Network unit tests

jonatack committed:
test: hoist proxy out of 2 network loops in feature_proxy.py

vasild reviewed (approved):
ACK d1d04469a96d15b4d0a30881cbe6760d916e0762

jonatack commented:
Note, this pull has a trivial one-line test conflict with #22834 that IMO should be reviewed/merged first so this pull doesn't invalidate the review there.

PastaPastaPasta commented on src/init.cpp:
```diff
@@ -1317,6 +1317,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         }
         for (int n = 0; n < NET_MAX; n++) {
             enum Network net = (enum Network)n;
```
please convert this c-style cast to a static_cast, since you are touching the line below
```suggestion
            enum Network net = static_cast<enum Network>(n);
```

jonatack replied:
Thanks, but this is unrelated to the changes here and I prefer to not touch it.

jonatack committed:
test: passing invalid -proxy raises expected init error

jonatack committed:
test: passing invalid -onion raises expected init error

jonatack committed:
test: passing -onlynet=onion without -proxy/-onion raises expected init error

jonatack committed:
test: passing -onlynet=onion with -onion=0/-noonion raises expected init error

jonatack commented:
Rebased following the merge of #22834 and added further test coverage and doc follow-ups to that pull.

vasild commented on test/functional/feature_proxy.py:
```diff
@@ -305,6 +308,32 @@ def networks_dict(d):
+        self.nodes[1].assert_start_raises_init_error(expected_msg=msg)
+
+        self.log.info("Test passing invalid -onion raises expected init error")
+        self.nodes[1].extra_args = ["-onion=xyz:123"]
+        msg = "Error: Invalid -onion address or hostname: 'xyz:123'"
```
`xyz:123` could be valid if `xyz` resolves. For example, `echo "127.0.0.1 xyz" >> /etc/hosts` bricks the test. Better use non-numeric port like in the `-proxy` test.

jonatack replied:
Done

jonatack replied:
(Good point!)

vasild commented on test/functional/feature_proxy.py:
```diff
@@ -263,12 +267,13 @@ def networks_dict(d):
         n2 = networks_dict(self.nodes[2].getnetworkinfo())
         assert_equal(NETWORKS, n2.keys())
+        proxy = f'{self.conf2.addr[0]}:{self.conf2.addr[1]}'
         for net in NETWORKS:
             if net == NET_I2P:
                 expected_proxy = ''
                 expected_randomize = False
             else:
-                expected_proxy = f'{self.conf2.addr[0]}:{self.conf2.addr[1]}'
+                expected_proxy = proxy
```
nit: this change looks unnecessary because the expression is used just once, so putting it in a variable does not make it more maintainable. It maybe makes it less readable because now one has to visually go up a few lines in order to find what `expected_proxy` ends up with.

jonatack replied:
This line is invoked 4 times in the loop, so it seems better to do the string interpolation before entering it.

vasild replied:
Oh, performance :eyes: :space_invader:

vasild commented on test/functional/feature_proxy.py:
```diff
@@ -279,11 +284,9 @@ def networks_dict(d):
             for net in NETWORKS:
-                if net == NET_I2P or net == NET_ONION:
-                    expected_proxy = ''
-                else:
-                    expected_proxy = f'[{self.conf3.addr[0]}]:{self.conf3.addr[1]}'
+                expected_proxy = '' if net == NET_I2P or net == NET_ONION else proxy
```
nit: this looks unnecessary coz the expression is used just once. Also, personally, I find the perl-style harder to read: `x=1 if some_condition else 2`.

I am fine as it is, but I will be happier if commit `test: hoist proxy out of 2 network loops in feature_proxy.py` is removed :)

jonatack replied:
This line is invoked 3 times in the loop, so it seems better to do the more complex lookups and string interpolation before entering it. The ternary is used frequently in our Python tests and I think it's more succinct here to use the ternary form for an assignment instead of if - else with two assignments.

vasild reviewed (commented):
Approach ACK

jonatack commented on src/init.cpp:
```diff
@@ -462,7 +462,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-i2psam=<ip:port>", "I2P SAM proxy to reach I2P peers and accept I2P connections (default: none)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-i2pacceptincoming", "If set and -i2psam is also set then incoming I2P connections are accepted via the SAM proxy. If this is not set but -i2psam is set then only outgoing connections will be made to the I2P network. Ignored if -i2psam is not set. Listening for incoming I2P connections is done through the SAM proxy, not by binding to a local address and port (default: 1)", Arg ...
-    argsman.AddArg("-onlynet=<net>", "Make automatic outgoing connections only through network <net> (" + Join(GetNetworkNames(), ", ") + "). Incoming connections are not affected by this option. This option can be specified multiple times to allow multiple networks.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-onlynet=<net>", "Make automatic outbound connections only through network <net> (" + Join(GetNetworkNames(), ", ") + "). Inbound and manual connections are not affected by this option. This option can be specified multiple times to allow multiple networks.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
I'm going to propose this documentation in another pull for potential backport to v23 if it doesn't make the branch-off (and drop it here); will look at adding your suggestion there.

jonatack replied:
> nit: "only through network"... hmm, what about making connections **to** IPv4 addresses **through** the Tor network? Could this be misinterpreted as "onlynet=onion will also make connections to IPv4 as long as it is through Tor"?
> 
> Is "only **to** network" better wording?

Agreed.

jonatack replied:
Done in #24468

vasild reviewed (approved):
ACK 58a14795b89a6bd812e0b71cb8b3088b8ab55c11

jonatack commented:
Dropped the documentation commit to propose it in a separate pull for v23 if this pull doesn't make the branch-off and updated one of the tests per https://github.com/bitcoin/bitcoin/pull/24205#discussion_r818633992 feedback by @vasild (thanks!)

brunoerg reviewed (approved):
ACK 58a14795b89a6bd812e0b71cb8b3088b8ab55c11

Checked the following init errors are now covered in functional tests:
https://github.com/bitcoin/bitcoin/blob/91d12344b1e51809c1ef6b630b631a6da00267c3/src/init.cpp#L1333
https://github.com/bitcoin/bitcoin/blob/91d12344b1e51809c1ef6b630b631a6da00267c3/src/init.cpp#L1352
https://github.com/bitcoin/bitcoin/blob/91d12344b1e51809c1ef6b630b631a6da00267c3/src/init.cpp#L1363

Code looks great for both tests. Just ran them several times and got no error.

jonatack commented:
I think this has been ready for a month now, if anyone would like to do a final review or merge.

dongcarl reviewed (approved):
Code Review ACK 58a14795b89a6bd812e0b71cb8b3088b8ab55c11

Change seems straightforward. Had to convince myself of the `assert` in `src/init.cpp`, but it seems right that we don't want users to specify `-onlynet=foo` and `foo` be somehow default disabled, and now `bitcoind` is silently operating without `foo`.

michaelfolkson commented:
Post-merge ACK 58a14795b89a6bd812e0b71cb8b3088b8ab55c11

Built on MacOS Big Sur, reviewed code changes and checked the functional test was covering the errors in the PR description.
